### PR TITLE
⚠️ Improve Context handling in clusterctl

### LINKED ISSUE #8733: Improve Context handling in clusterctl
### What would you like to be added (User Story)?

Today we have plenty of places in clusterctl where we use context.TODO(). This issue is about incrementally improving clusterctl to pass through Context top-down.

Related:
* The recent CR bump PR (https://github.com/kubernetes-sigs/cluster-api/pull/8007) introduced the new wait utils from apimachinery which means it's now even easier to pass through context into code that is run inside of the wait utils.

### Detailed Description

-

### Anything else you would like to add?

_No response_

### Label(s) to be applied

/kind cleanup
/area clusterctl

One or more /area label. See https://github.com/kubernetes-sigs/cluster-api/labels?q=area for the list of labels.

### PR DESCRIPTION
**Pull request by Fedosin**
**What this PR does / why we need it**:
Today we have plenty of places in `clusterctl` where we use context.TODO(). Moreover in many places `ctx` is defined as a global variable, which is considered as an anti-pattern. This PR allows to pass context variable to all public functions to manage their behavior.

**Note:** This code breaks backward compatibility as it changes public function signatures by adding `context` parameter to them.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #8733

### CODE REVIEW

**Comment by enxebre** [comment]
Overall change lgtm. This is affecting multiple public functions so we might want to consider a deprecation notice, not sure what'd best way given the surface of the change.
Given the size of the PR, splitting this into multiple PRs would probably accelerate the review process.

**Comment by sbueringer** [comment]
DIdn't have time to review. Just a quick note about API changes in clusterctl:


> When this package is used as a library, we do not currently provide any compatibility guarantees. We will make reasonable efforts to follow a typical deprecation period prior to removal, but breaking changes can happen when necessary.
> https://github.com/kubernetes-sigs/cluster-api/tree/main/cmd/clusterctl

**Comment by fabriziopandini** [comment]
It will take some time for me to review such a big PR; it would be great if clusterctl reviewers can make a preliminary pass first 

cc @ykakarap @Jont828 

>  I think capi operator is the main consumer of clusterctl functions

We have many other users that have built tooling or automation on top of clusterctl functions; in fact clusterctl has been built as a library for those use cases, and the operator was added on top later.

Considering this, even if we are allowed to do such changes, given that they are so extensive I think we should advertise them a little bit and let the community provide feedback before merging at this stage of the release.

**Comment by ykakarap** [comment]
I can take a pass at this.

**Comment by Fedosin** [comment]
@ykakarap thanks! ping me here or in slack if you have any questions about this PR.

File: cmd/clusterctl/client/config/client.go [hunk 1]
```diff
@@ -102,16 +104,16 @@ func newConfigClient(path string, options ...Option) (*configClient, error) {
 // Reader define the behaviours of a configuration reader.
 type Reader interface {
 	// Init allows to initialize the configuration reader.
-	Init(path string) error
+	Init(ctx context.Context, path string) error
 
 	// Get returns a configuration value of type string.
 	// In case the configuration value does not exists, it returns an error.
-	Get(key string) (string, error)
+	Get(ctx context.Context, key string) (string, error)
```
Comment (ykakarap) [inline cmd/clusterctl/client/config/client.go:111]: Looks like this ctx is not used in any of the implmentations, isn't it? Can you check if this is safe to drop?

Comment (ykakarap) [reply cmd/clusterctl/client/config/client.go:111]: With that I think you can slo drop `ctx` from `UnmarshalKey`, isn't it?

File: cmd/clusterctl/client/config_test.go [hunk 2]
```diff
@@ -849,9 +866,9 @@ func newFakeClientWithoutCluster(configClient config.Client) *fakeClient {
 	}
 
 	var err error
-	fake.internalClient, err = newClusterctlClient("fake-config",
+	fake.internalClient, err = newClusterctlClient(context.TODO(), "fake-config",
```
Comment (ykakarap) [inline cmd/clusterctl/client/config_test.go:869]:
```suggestion
	fake.internalClient, err = newClusterctlClient(context.Background(), "fake-config",
```

Should this be `context.Background()`?

**Review by ykakarap — COMMENTED** [review]
Overall LGTM.

**Comment by fabriziopandini** [comment]
/hold 
for community feedback

File: cmd/clusterctl/client/config/client.go (hunk 1 above)
Comment (Fedosin) [reply cmd/clusterctl/client/config/client.go]: Yes, plus I also updated `AlterImage` and `NewTemplate`.

File: cmd/clusterctl/client/cluster/cert_manager.go [hunk 3]
```diff
@@ -162,10 +162,10 @@ func (cm *certManagerClient) EnsureInstalled() error {
 	// Otherwise install cert manager.
 	// NOTE: this instance of cert-manager will have clusterctl specific annotations that will be used to
 	// manage the lifecycle of all the components.
-	return cm.install()
+	return cm.install(ctx)
 }
 
-func (cm *certManagerClient) install() error {
+func (cm *certManagerClient) install(ctx context.Context) error {
```
Comment (Jont828) [inline cmd/clusterctl/client/cluster/cert_manager.go:168]: In the line below, can we get a logger from the context now that we're passing one in? I'm not sure as we're not in a controller.

**Review by Jont828 — COMMENTED** [review]
Nice work! This must have taken a while. Everything looks pretty straightforward to me. Just had a question about if we can use the context we're passing in for logs.

File: cmd/clusterctl/client/cluster/cert_manager.go (hunk 3 above)
Comment (Jont828) [reply cmd/clusterctl/client/cluster/cert_manager.go:168]: That's fine by me, this PR is already pretty large, and we can consider a follow up if we still want a log from context.

**Comment by fabriziopandini** [comment]
@Fedosin I propose to move forward with this PR by changing the PR type into breaking change and adding a note about this change into https://main.cluster-api.sigs.k8s.io/developer/providers/migrations/v1.5-to-v1.6

@sbueringer @killianmuldoon opinions?

**Comment by sbueringer** [comment]
Thank you very much!!

/lgtm
/approve

/hold
for e2e-full

I'm planning to merge this PR as soon as CI is green to avoid further rebasing (we also have a few more huge PRs in the pipeline)

@Fedosin Please follow-up with a PR to "add a note about this change into [main.cluster-api.sigs.k8s.io/developer/providers/migrations/v1.5-to-v1.6](https://main.cluster-api.sigs.k8s.io/developer/providers/migrations/v1.5-to-v1.6)"

**Comment by sbueringer** [comment]
Thank you very much!!

/lgtm

Can you squash the commits (with just a squash without another rebase onto main the lgtm label should be preserved)

Feel free to hold cancel after squash + e2e-full green

**Comment by Fedosin** [comment]
thank you all so much for the review!

/hold cancel

**Comment by sbueringer** [comment]
Thx, especially for the patience! :)

**Pull request merged**